### PR TITLE
[Macros] Diagnose macro expansions containing invalid declarations.

### PR DESCRIPTION
**Pull request by hborla**
Diagnose macro expansions that produce any of the following declarations:
 - Import declarations
 - Operator and precedence group declarations
 - Macro declarations
 - Extensions
 - Types with `@main` attributes
 - Top-level default literal type overrides
 - Value decls with names not covered by the macro declaration.

Resolves: rdar://105022807

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/Sema/TypeCheckMacros.cpp [hunk 1]
```diff
@@ -476,6 +476,104 @@ ExpandPeerMacroRequest::evaluate(Evaluator &evaluator, Decl *decl) const {
...
+
+    // Diagnose value decls with names not covered by the macro
+    if (auto *value = dyn_cast<ValueDecl>(decl)) {
+      auto baseName = value->getBaseName();
+      if (baseName.isSpecial()) {
+        baseName = ctx.getIdentifier(baseName.userFacingName());
+      }
+
+      // $-prefixed names are unique names. These are always allowed.
+      if (baseName.getIdentifier().hasDollarPrefix()) {
```
Comment (DougGregor) [inline lib/Sema/TypeCheckMacros.cpp:562]: It's fine for it to be a follow-up, but I think that when there is a dollar prefix, we should go demangle the name and make sure it's one of our unique names.

**Review by DougGregor — APPROVED** [review]
This looks fantastic, thank you!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
https://github.com/apple/swift-syntax/pull/1387

@swift-ci please smoke test macOS

**Pull request merged**